RecoBTag/ONNXRuntime/plugins/UnifiedParticleTransformerAK4ONNXJetTagsProducer.cc

```diff
@@ -0,0 +1,368 @@
+    *(++ptr) = lt_features.drminsv;
+    *(++ptr) = lt_features.charge;
+    *(++ptr) = lt_features.puppiw;
+    *(++ptr) = lt_features.chi2;
+    *(++ptr) = lt_features.quality;
+    *(++ptr) = lt_features.lostInnerHits;
+    *(++ptr) = lt_features.numberOfPixelHits;
+    *(++ptr) = lt_features.numberOfStripHits;
+    assert(start + n_features_lt_ - 1 == ptr);
+  }
+
+  // n_pf candidates
+  auto max_n_pf_n = std::min(features.n_pf_features.size(), (std::size_t)n_npf_);
+  for (std::size_t n_pf_n = 0; n_pf_n < max_n_pf_n; n_pf_n++) {
+    const auto& n_pf_features = features.n_pf_features.at(n_pf_n);
+    ptr = &data_[kNeutralCandidates][offset + n_pf_n * n_features_npf_];
+    start = ptr;
+    *ptr = n_pf_features.ptrel;
+    *(++ptr) = n_pf_features.etarel;
+    *(++ptr) = n_pf_features.phirel;
+    *(++ptr) = n_pf_features.deltaR;
+    *(++ptr) = n_pf_features.isGamma;
+    *(++ptr) = n_pf_features.hadFrac;
+    *(++ptr) = n_pf_features.drminsv;
+    *(++ptr) = n_pf_features.puppiw;
+    assert(start + n_features_npf_ - 1 == ptr);
+  }
+
+  // sv candidates
+  auto max_sv_n = std::min(features.sv_features.size(), (std::size_t)n_sv_);
+  for (std::size_t sv_n = 0; sv_n < max_sv_n; sv_n++) {
+    const auto& sv_features = features.sv_features.at(sv_n);
+    ptr = &data_[kVertices][offset + sv_n * n_features_sv_];
+    start = ptr;
+    *ptr = sv_features.pt;
+    *(++ptr) = sv_features.deltaR;
+    *(++ptr) = sv_features.mass;
+    *(++ptr) = sv_features.etarel;
+    *(++ptr) = sv_features.phirel;
+    *(++ptr) = sv_features.ntracks;
+    *(++ptr) = sv_features.chi2;
+    *(++ptr) = sv_features.normchi2;
+    *(++ptr) = sv_features.dxy;
+    *(++ptr) = sv_features.dxysig;
+    *(++ptr) = sv_features.d3d;
+    *(++ptr) = sv_features.d3dsig;
+    *(++ptr) = sv_features.costhetasvpv;
+    *(++ptr) = sv_features.enratio;
+    assert(start + n_features_sv_ - 1 == ptr);
+  }
+
+  // cpf pairwise features (4-vectors)
+  auto max_cpf_n = std::min(features.c_pf_features.size(), (std::size_t)n_cpf_);
+  for (std::size_t cpf_n = 0; cpf_n < max_cpf_n; cpf_n++) {
+    const auto& cpf_pairwise_features = features.c_pf_features.at(cpf_n);
+    ptr = &data_[kChargedCandidates4Vec][offset + cpf_n * n_pairwise_features_cpf_];
+    start = ptr;
+    *ptr = cpf_pairwise_features.px;
+    *(++ptr) = cpf_pairwise_features.py;
+    *(++ptr) = cpf_pairwise_features.pz;
+    *(++ptr) = cpf_pairwise_features.e;
+
+    assert(start + n_pairwise_features_cpf_ - 1 == ptr);
+  }
+
+  // lt pairwise features (4-vectors) specific case requiring (pt,eta,phi,e)
+  auto max_lt_N = std::min(features.lt_features.size(), (std::size_t)n_lt_);
+  for (std::size_t lt_N = 0; lt_N < max_lt_N; lt_N++) {
+    const auto& lt_pairwise_features = features.lt_features.at(lt_N);
+    ptr = &data_[kLostTracks4Vec][offset + lt_N * n_pairwise_features_lt_];
+    start = ptr;
+    *ptr = lt_pairwise_features.pt;
+    *(++ptr) = lt_pairwise_features.eta;
```

Comment:
Why pt/eta/phi/e is used here, while all the others are using px/py/pz/e?

Reply:
This is intended, we have a filter inside the model for rejecting any outlier if the lost track production encounters some. For this we use this set for the 4 momentum, After the filter, the lost tracks pairwise features are converted to the px,py,pz,e for matching the same formalism as the other jet constituents